poseidon-sponge.md

```diff
@@ -0,0 +1,240 @@
+# Poseidon sponge construction
+
+- Version: 0.0.0
+- Maintainer: Jack Grigg
+
+## Introduction
+
+This document specifies the Poseidon sponge construction, a duplex sponge [[BDPV2011]]
+instantiated with the [Poseidon permutation]. It also specifies several instantiations of
+the sponge for use in various hashing domains.
+
+[BDPV2011]: https://keccak.team/files/SpongeDuplex.pdf
+[Poseidon permutation]: ./poseidon-permutation.md
+
+## Notation and conventions
+
+- `[a, b)` means the sequence of integers from `a` inclusive to `b` exclusive.
+- `[x; N]` means an array of length `N` where every entry is filled with `x`.
+- `y[i]` means the element of the array `y` at index `i`.
+- `len(y)` means the length of the array `y`. For example, `len([x; N]) == N`.
+
+## Duplex sponge
+
+The duplex sponge `PoseidonSponge` is built around `PoseidonPermutation`, a permutation of
+width `T`, which encapsulates the specific Poseidon parameters for the sponge. It has a
+parameter `RATE`, the number of input field elements processed by each invocation of the
+permutation, which is at most `T - 1`. The value `C = T - RATE` is called the capacity.
+
+> TODO: Describe how the permutation instance and the rate relate to each other and to the
+> desired security parameter `M`.
+
+`PoseidonSponge` has several internal properties:
+- An internal state of `T` field elements.
+- A mode variable that is either `absorbing` or `squeezing`.
+- A cache of field elements corresponding to the current mode.
+- A duplex padding function that is applied to input elements prior to incorporating them
+  into the sponge state.
+
+The only operations exposed by the sponge are:
+- `PoseidonSponge.Initialize`, which prepares a new sponge.
+- `PoseidonSponge.Absorb`, which takes a single field element into the sponge.
+- `PoseidonSponge.Squeeze`, which produces a single field element derived from all prior
+  `Absorb` and `Squeeze` operations on the sponge, as well as the initialization values.
```

Comment:
> I suggest having the arguments here and dropping the `PoseidonSponge` in front as this seems language specific.

Heh, this was my attempt at writing the spec in a language-agnostic way! If having function names without periods is less likely to cause someone to assume a particular language, we can do that. I note that defining it in the way @rozbb suggests would require the reader to have at least some familiarity with Rust specifically.

> "Duplex sponge" isn't really a term used anywhere in the literature. They call it a duplex.

That's fair. However, what is specified here is neither solely a duplex, nor solely a sponge - it's a hybrid built around a duplex construction. In the current spec, `PoseidonSponge.PerformDuplex` is the function that implements the duplex construction, while `PoseidonSponge` itself can be used as a duplex ("absorb RATE, squeeze RATE, repeat), as a sponge (absorb many elements, squeeze many elements, throw the sponge away), or as something in between (e.g. a transcript like in STROBE). Is there a better name for that general construction than "sponge"?

> if we're defining high-level functions, why not have an absorb function that takes a vector of field elements?

Sure, we could define that. I figured it was a trivial wrapper around the single-element absorb function, but we could make it explicit.

Reply:
> Is there a better name for that general construction than "sponge"?

I think, as I just pointed below, that it's going to be hard to specify a hybrid of a sponge and a duplex because they have different requirements (specifically the compliant padding that must occur before every permutation in the duplex).

Reply:
My understanding is that `PoseidonSponge` is a construction built on a duplex, which is build on a sponge.

What this new construction (which I agree, should have another name. Maybe "Extendable Duplex"?) adds on top of a duplex is supporting arbitrary-sized inputs and outputs. Each duplex call accepts inputs and outputs whose size is <= RATE (after padding). The new construction allows arbitrary-sized inputs by splitting it into chunks of size RATE (after padding) and calling the duplex multiple times discarding the outputs. And then allows arbitrary-sized outputs by calling the duplex multiple times with null inputs (which will still be padded...). Is that correct? If so, my suggestion is to actually define the construction on top of the duplex. The description below would then be an implementation guide (which is important, but doesn't seem to be the ideal way of actually defining the construction).

Apologies if I'm not making sense, I've just read the sponge and duplex papers, but hopefully I can contribute somehow with a fresh perspective.